hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -1336,6 +1336,132 @@ case class TableKeyByAndAggregate(
+        }
+
+        val newRowType = localKeyPType ++ rTyp
+
+        val localBufferSize = bufferSize
+        val rdd = prev.rvd
+          .boundary
+          .mapPartitionsWithIndex { (i, ctx, it) =>
+            val partRegion = ctx.freshRegion
+            val globals = globalsBc.value.readRegionValue(partRegion)
+            val makeKey = {
+              val f = makeKeyF(i, partRegion)
+              rv: RegionValue => {
+                val keyOff = f(rv.region, rv.offset, false, globals, false)
+                SafeRow.read(localKeyPType, rv.region, keyOff).asInstanceOf[Row]
+              }
+            }
+            val makeAgg = { () =>
+              val aggRegion = ctx.freshRegion
+              RegionValue(aggRegion, deserialize(aggRegion, initAggs))
+            }
+
+            val seqOp = {
+              val f = makeSeq(i, partRegion)
+              (rv: RegionValue, agg: RegionValue) => {
+                f.setAggState(agg.region, agg.offset)
+                f(rv.region, globals, false, rv.offset, false)
+                agg.setOffset(f.getAggOffset())
+              }
+            }
+            val serializeAndCleanupAggs = { rv: RegionValue =>
+              val a = serialize(rv.region, rv.offset)
+              rv.region.close()
+              a
+            }
+
+            new BufferedAggregatorIterator[RegionValue, RegionValue, Array[Byte], Row](
+              it,
+              makeAgg,
+              makeKey,
+              seqOp,
+              serializeAndCleanupAggs,
+              localBufferSize)
+          }.aggregateByKey(initAggs, nPartitions.getOrElse(prev.rvd.getNumPartitions))(combOp, combOp)
+
+        val crdd = ContextRDD.weaken(rdd).cmapPartitionsWithIndex(
+          { (i, ctx, it) =>
+            val region = ctx.region
+
+            val rvb = new RegionValueBuilder()
+            val partRegion = ctx.freshRegion
+            val globals = globalsBc.value.readRegionValue(partRegion)
+            val annotate = makeAnnotate(i, partRegion)
+
+            val rv = RegionValue(region)
+            it.map { case (key, aggs) =>
+              rvb.set(region)
+              rvb.start(newRowType)
+              rvb.startStruct()
+              var i = 0
+              while (i < localKeyType.size) {
+                rvb.addAnnotation(localKeyType.types(i), key.get(i))
+                i += 1
+              }
+
+              val aggOff = deserialize(rv.region, aggs)
+              annotate.setAggState(rv.region, aggOff)
+              rvb.addAllFields(rTyp, region, annotate(region, globals, false))
+              rvb.endStruct()
+              rv.setOffset(rvb.end())
+              rv
+            }
+          })
+
+        return prev.copy(
+          typ = typ,
+          rvd = RVD.coerce(RVDType(newRowType, keyType.fieldNames), crdd))
+      } catch {
+        case e: agg.UnsupportedExtraction =>
+          log.info(s"couldn't lower TableAggregate: $e")
```

Comment:
fix

Reply:
just need to fix the copy/paste error here and we can merge. Can put in benchmarks separately.